src/main/java/net/kaczmarzyk/spring/data/jpa/domain/JoinFetch.java

```diff
@@ -31,69 +32,114 @@
-        this.distinct = distinct;
-    }
-
-    @Override
-    public Predicate toPredicate(Root<T> root, CriteriaQuery<?> query, CriteriaBuilder cb) {
-        query.distinct(distinct);
-        if (!Number.class.isAssignableFrom(query.getResultType())) { // do not join in count queries
-            for (String path : pathsToFetch){
-                root.fetch(path, joinType);
-            }
-        }
-        return null;
-    }
-
-    @Override
-    public int hashCode() {
-        final int prime = 31;
-        int result = 1;
-        result = prime * result + ((joinType == null) ? 0 : joinType.hashCode());
-        result = prime * result + ((pathsToFetch == null) ? 0 : pathsToFetch.hashCode());
-        return result;
-    }
-
-    @Override
-    public boolean equals(Object obj) {
-        if (this == obj) {
-            return true;
-        }
-        if (obj == null) {
-            return false;
-        }
-        if (getClass() != obj.getClass()) {
-            return false;
-        }
-        @SuppressWarnings("rawtypes")
-        JoinFetch other = (JoinFetch) obj;
-        if (joinType != other.joinType) {
-            return false;
-        }
-        if (pathsToFetch == null) {
-            if (other.pathsToFetch != null) {
-                return false;
-            }
-        } else if (!pathsToFetch.equals(other.pathsToFetch)) {
-            return false;
-        }
-        return true;
-    }
-
-    @Override
-    public String toString() {
-        return "JoinFetch[" +
-                "pathsToFetch=" + pathsToFetch +
-                ", joinType=" + joinType +
-                ']';
-    }
+
+	public JoinFetch(QueryContext queryContext, String[] pathsToFetch, JoinType joinType, boolean distinct) {
+		this(queryContext, pathsToFetch, "", joinType, distinct);
+	}
+
+	public JoinFetch(QueryContext queryContext, String[] pathsToFetch, String alias, JoinType joinType, boolean distinct) {
+		this.context = queryContext;
+		this.pathsToFetch = Arrays.asList(pathsToFetch);
+		this.alias = alias;
+		this.joinType = joinType;
+		this.distinct = distinct;
+
+		if (!alias.isEmpty() && pathsToFetch.length != 1) {
+			throw new IllegalArgumentException("Join fetch alias can be defined for join fetch with single path!");
+		}
+	}
+
+	@Override
+	public Predicate toPredicate(Root<T> root, CriteriaQuery<?> query, CriteriaBuilder cb) {
+		query.distinct(distinct);
+		if (!Number.class.isAssignableFrom(query.getResultType())) { // do not join in count queries
+			if (pathsToFetch.size() == 1) {
+				String pathToFetch = pathsToFetch.get(0);
+				if (pathToJoinContainsAlias(pathToFetch)) {
+					String[] pathToJoinFetchOnSplittedByDot = pathToJoinFetchSplittedByDto(pathToFetch);
+					String alias = pathToJoinFetchOnSplittedByDot[0];
+					String path = pathToJoinFetchOnSplittedByDot[1];
+
+					Fetch<?, ?> evaluated = context.getEvaluatedJoinFetch(alias);
+					evaluated = evaluated.fetch(path, joinType);
```

Comment:
Consider not reusing "evaluated" variable for 2 purposes, it's a bit uncler for a new reader